File: cpp/589.N-ary Tree Preorder Traversal.cpp
/**
 * Source: LeetCode 589. N-ary Tree Preorder Traversal
 * Author: ceezyyy
 * Date: 10-16-2019
 */


/*
Given an n - ary tree, return the preorder traversal of its nodes' values.

For example, given a 3 - ary tree :

       1
     / / \
	3 2   4
   / \
  5   6

Return its preorder traversal as : [1, 3, 5, 6, 2, 4].

 */


#include<iostream>
#include<vector>
using namespace std;

// Definition for a Node.
class Node {
public:
	int val;
	vector<Node*> children;

	Node() {}

	Node(int _val, vector<Node*> _children) {
		val = _val;
		children = _children;
	}
};

class Solution {
public:
	vector<int> res;
	vector<int> preorder(Node* root) {
		if (!root)  // alway the first step in 'tree'
			return res;
		DFS(root);
		return res;  // do not forget to return value
	}

	void DFS(Node* node) {
		if (!node)
			return;
		res.push_back(node->val);
		for (Node* temp : node->children) {  // each children
			DFS(temp);  // recursion
		}
	}
};


/**
 * Runtime: 156 ms, faster than 44.49% of C++ online submissions for N-ary Tree Preorder Traversal.
 * Memory Usage: 32.5 MB, less than 94.74% of C++ online submissions for N-ary Tree Preorder Traversal.
 */

File: Python3/589.N-ary Tree Preorder Traversal.py
"""
@Title: 589. N-ary Tree Preorder Traversal
@Tag: N-ary tree
@Date: Feb-06 2020
@Author: ceezyyy
@Difficulty: Easy
"""


"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if not root:
            return []
        stack, res = [root], []
        while stack:
            node = stack.pop()
            res.append(node.val)
            stack.extend([child for child in node.children][::-1])
        return res


"""
Runtime: 52 ms, faster than 58.70% of Python3 online submissions for N-ary Tree Preorder Traversal.
Memory Usage: 14.7 MB, less than 100.00% of Python3 online submissions for N-ary Tree Preorder Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
